### PR TITLE
Gutenberg Integration: Make Authenticated Requests to the v2 REST API

### PR DESCRIPTION
**Pull request by rodrigoi**
Part of #26639

The PR adds a new middleware for `apiFetch` that routes all the API throughout the WordPress.com REST Proxy. This will not work on the Desktop Application until `oauthToken` detection is implemented, so only test on web.

To test this you'll have to:

* Enable debug logs by running `localStorage.setItem( 'debug', 'calypso:gutenberg' );` on your browser's console.
* Browse `calypso.localhost:3000/gutenberg/post/{site_slug}`.
* Look for `calypso:gutenberg sending API request to:` debug messages on the console.
* Verify that for each of the debug messages, there is a corresponding network request that returns correctly.

Pending Items

* Validate that the API responses are stored correctly and reflected on the UI
* Validate that there are no edge cases for `/wp/v2/` -> `/wp/v2/sites/{siteSlug}/`

### CODE REVIEW

File: client/gutenberg/editor/main.jsx [hunk 1]
```diff
@@ -36,7 +36,7 @@ class GutenbergEditor extends Component {
 			return null;
 		}
 
-		overrideAPIPaths( this.props.siteSlug );
+		applyMiddlewares( this.props.siteSlug );
```
Comment (vindl) [inline client/gutenberg/editor/main.jsx]: :+1: for the rename, maybe `applyAPIMiddlewares` would be more appropriate?

File: client/gutenberg/editor/utils.js [hunk 2]
```diff
@@ -12,17 +12,33 @@ import debugFactory from 'debug';
...
 
-export const overrideAPIPaths = siteSlug => {
-	// TODO: no API support for now. We'll also need to handle authorization here.
-	apiFetch.use( () => noop );
+export const applyMiddlewares = siteSlug => {
+	//make authenticated calls using the WordPress.com REST Proxy
+	//bypassing the apiFetch call that uses window.fetch
+	//first middleware in, last out
+	apiFetch.use( options => {
+		return proxy(
```
Comment (vindl) [inline client/gutenberg/editor/utils.js]: Optional: I think naming this `wpcomProxyRequest` would make it slightly more readable.

File: client/gutenberg/editor/utils.js [hunk 3]
```diff
@@ -12,17 +12,33 @@ import debugFactory from 'debug';
...
+	} );
 
 	const rootURL = 'https://public-api.wordpress.com/';
 	apiFetch.use( apiFetch.createRootURLMiddleware( rootURL ) );
 
 	// rewrite default API paths to match WP.com equivalents
 	// Example: /wp/v2/posts -> /wp/v2/sites/{siteSlug}/posts
 	apiFetch.use( ( options, next ) => {
-		const wpcomPath = `/wp/v2/sites/${ siteSlug }/` + options.path.replace( '/wp/v2/', '' );
+		const wpcomPath = `/sites/${ siteSlug }` + options.path.replace( '/wp/v2/', '/' );
```
Comment (vindl) [inline client/gutenberg/editor/utils.js]: We could probably shorten this to:
```javascript
const wpcomPath = options.path.replace( '/wp/v2/', `/sites/${siteSlug}/` );
```

Comment (rodrigoi) [reply client/gutenberg/editor/utils.js]: Very wise. Updated on fca7b6135a91b47113af39793ebf0812a14bb32b.

File: client/gutenberg/editor/utils.js (hunk 2 above)
Comment (rodrigoi) [reply client/gutenberg/editor/utils.js]: I should write 1000 times "do not copy and paste from the examples". Updated on fca7b6135a91b47113af39793ebf0812a14bb32b

File: client/gutenberg/editor/main.jsx (hunk 1 above)
Comment (rodrigoi) [reply client/gutenberg/editor/main.jsx]: I was thinking on dropping the `utils.js` in favor of `middlewares.js` or something like that, but since naming things is hard, I'll leave it for another PR. Updated on fca7b6135a91b47113af39793ebf0812a14bb32b.

File: client/gutenberg/editor/utils.js [hunk 4]
```diff
@@ -12,17 +12,33 @@ import debugFactory from 'debug';
...
+				apiNamespace: 'wp/v2',
+			},
+			( error, body ) => {
+				if ( error ) {
+					return Promise.reject( error );
+				}
+
+				return Promise.resolve( body );
+			}
+		);
```
Comment (youknowriad) [inline client/gutenberg/editor/utils.js]: apiFetch is supposed to return the response object if `parse=false` in the options. (We use the `window.fetch` response object shape as a public API), so if we want to be 100% compatible here, we need to support that case too. 

It's probably used on for `metaboxes` right now in Gutenberg and Metaboxes are not used in Calypso so it should work without it but that doesn't guarantee that there won't be any usage of the `parse` flag outside metaboxes in the future. (I think I saw a PR not merged yet using it)

**Comment by dmsnell** [comment]
Are we unable to use the data layer here for these? Since this is already only running in Calypso it'd be great if instead of using the proxy request `wpcom.js` object itself we could jump into the network cod we already have which centrally provides things like the auto-retry, request deduplication, inspection, and other policies.

That would be…instead of making a direct network call, translating the `apiFetch` command into a Redux action to dispatch - `WPCOM_HTTP_REQUEST` with the appropriate parameters.

What is your assessment of this other approach?

File: client/gutenberg/editor/utils.js [hunk 5]
```diff
@@ -12,17 +12,35 @@ import debugFactory from 'debug';
...
 const debug = debugFactory( 'calypso:gutenberg' );
 
-export const overrideAPIPaths = siteSlug => {
-	// TODO: no API support for now. We'll also need to handle authorization here.
-	apiFetch.use( () => noop );
+export const applyAPIMiddlewares = siteSlug => {
+	//make authenticated calls using the WordPress.com REST Proxy
+	//bypassing the apiFetch call that uses window.fetch
+	//first middleware in, last out
+	apiFetch.use( options => {
```
Comment (gwwar) [inline client/gutenberg/editor/utils.js]: Is it a problem that we never call `next` here? This likely breaks the middleware chain right?

**Comment by gwwar** [comment]
> This will not work on the Desktop Application until oauthToken detection is implemented, so only test on web.

I believe it should be sufficient to check for the `oauth` feature flag, but it's fine to hold off support until later.

When we get to that point, @v18 had some excellent testing instructions in https://github.com/Automattic/wp-calypso/pull/8456 that we can reuse for setting up oauth for local envs in case we needed a faster dev loop than wp-desktop

**Comment by rodrigoi** [comment]
@dmsnell I'd like to follow up on your idea of using the data layer on a separate janitorial PR to avoid blocking other integration efforts 😉
